usr/netlink.c

```diff
@@ -443,7 +443,7 @@ kdestroy_session(uint64_t transport_handle, uint32_t sid)
 
 	memset(&ev, 0, sizeof(struct iscsi_uevent));
 
-	ev.type = ISCSI_UEVENT_DESTROY_SESSION;
+	ev.type = ISCSI_UEVENT_DESTROY_SESSION_ASYNC;
 	ev.transport_handle = transport_handle;
 	ev.u.d_session.sid = sid;
 
```

Comment:
It's going to be more complicated then just switching the type.

1. We need to support older kernels that didn't have this cmd. So you would try async first. If that returns a not supported error then drop down to sync.
2. For async iscsid can't free eveything once the cmd returns from the kernel, or we will have a case where the kernel still thinks there is a session, but userspace does not. A user might then expect his LUNs and session are not in use and do other management ops like umount/reuse a LUN or relogin. They will then get errors but won't know why since iscsiadm's logout command returned success. We have to wait for the kernel to report the session is destroyed with this event in drivers/scsi/scsi_transport_iscsi.c:

void iscsi_free_session(struct iscsi_cls_session *session)
{
        ISCSI_DBG_TRANS_SESSION(session, "Freeing session\n");
        iscsi_session_event(session, ISCSI_KEVENT_DESTROY_SESSION);
        put_device(&session->dev);
}

In open-iscsi we just use this for this one weird driver, qla4xxx, in open-iscsi/user/initiator.c:iscsi_async_session_destruction and all we do is log a debug message. That driver works in a way where all the state is stored in the card's flash. So instead of have the node DB. It has a DB  of nodes in its flash that's read/written to, and open-iscsi just gets/sets that stuff.

For the non-qla4xxx case we have to modify the code so iscsi_async_session_destruction frees the userspace/iscsid session at that time.

So it will work more like the target/session unbind operation where session_unbind sends a NL cmd to tell the kernel to delete the iscsi target and it's LUNs from the session. The kernel returns immediately, and then does a:

        iscsi_session_event(session, ISCSI_KEVENT_UNBIND_SESSION);

to tell userspace when the target is deleted. open-iscsi then gets that event and proceeds to stop and teardown the conn/session.

Reply:
Hi @mikechristie, thanks for your review

I want to do this as following:

1. add a new callback destroy_session_async in iscsi_ipc, which implements a ISCSI_UEVENT_DESTROY_SESSION_ASYNC, like following:

```
diff --git a/include/iscsi_if.h b/include/iscsi_if.h
index 22b7c09..50f3b1b 100644
--- a/include/iscsi_if.h
+++ b/include/iscsi_if.h
@@ -82,6 +82,7 @@ enum iscsi_uevent_e {
        ISCSI_UEVENT_LOGOUT_FLASHNODE_SID       = UEVENT_BASE + 30,
        ISCSI_UEVENT_SET_CHAP           = UEVENT_BASE + 31,
        ISCSI_UEVENT_GET_HOST_STATS     = UEVENT_BASE + 32,
+       ISCSI_UEVENT_DESTROY_SESSION_ASYNC = UEVENT_BASE +33,
        ISCSI_UEVENT_MAX                = ISCSI_UEVENT_GET_HOST_STATS,
 
        /* up events */
diff --git a/usr/iscsi_ipc.h b/usr/iscsi_ipc.h
index 47857dd..99e236f 100644
--- a/usr/iscsi_ipc.h
+++ b/usr/iscsi_ipc.h
@@ -78,6 +78,7 @@ struct iscsi_ipc {
                               uint32_t *hostno);
 
        int (*destroy_session) (uint64_t transport_handle, uint32_t sid);
+       int (*destroy_session_async) (uint64_t transport_handle, uint32_t sid);
 
        int (*unbind_session) (uint64_t transport_handle, uint32_t sid);
 
diff --git a/usr/netlink.c b/usr/netlink.c
index 22cad83..04edccc 100644
--- a/usr/netlink.c
+++ b/usr/netlink.c
@@ -456,6 +456,30 @@ kdestroy_session(uint64_t transport_handle, uint32_t sid)
        return 0;
 }
 
+static int
+kdestroy_session_async(uint64_t transport_handle, uint32_t sid)
+{
+       int rc;
+       struct iscsi_uevent ev;
+       struct iovec iov[2];
+
+       log_debug(7, "in %s", __FUNCTION__);
+
+       memset(&ev, 0, sizeof(struct iscsi_uevent));
+
+       ev.type = ISCSI_UEVENT_DESTROY_SESSION_ASYNC;
+       ev.transport_handle = transport_handle;
+       ev.u.d_session.sid = sid;
+
+       iov[1].iov_base = &ev;
+       iov[1].iov_len = sizeof(ev);
+       rc = __kipc_call(iov, 2);
+       if (rc < 0)
+               return rc;
+
+       return 0;
+}
+
 static int
 kunbind_session(uint64_t transport_handle, uint32_t sid)
 {
@@ -1763,6 +1787,7 @@ struct iscsi_ipc nl_ipc = {
        .sendtargets            = ksendtargets,
        .create_session         = kcreate_session,
        .destroy_session        = kdestroy_session,
+       .destroy_session_async  = kdestroy_session_async,
        .unbind_session         = kunbind_session,
        .create_conn            = kcreate_conn,
        .destroy_conn           = kdestroy_conn,
```

2. in session_conn_shutdown(), try destroy_session_async first:

```
diff --git a/usr/initiator.c b/usr/initiator.c
index 7186729..98b07a8 100644
--- a/usr/initiator.c
+++ b/usr/initiator.c
@@ -498,15 +498,24 @@ session_conn_shutdown(iscsi_conn_t *conn, queue_task_t *qtask,
 
 cleanup:
        if (session->id != INVALID_SESSION_ID) {
+               int err = 0;
                log_debug(2, "kdestroy session %u", session->id);
                session->r_stage = R_STAGE_SESSION_DESTOYED;
-               if (ipc->destroy_session(session->t->handle, session->id)) {
-                       log_error("can not safely destroy session %d",
-                                 session->id);
-                       return ISCSI_ERR_INTERNAL;
+               err = ipc->destroy_session_async(session->t->handle, session->id);
+               if (err == -ENOSYS) {
+                       err = ipc->destroy_session(session->t->handle, session->id);
+                       if (err) {
+                               log_error("can not safely destroy session %d",
+                                       session->id);
+                               return ISCSI_ERR_INTERNAL;
+                       } else {
+                               goto out;
+                       }
                }
+               return err ? ISCSI_ERR_INTERNAL : ISCSI_SUCCESS;
        }
 
+out:
        log_warning("Connection%d:%d to [target: %s, portal: %s,%d] "
                    "through [iface: %s] is shutdown.",
                    session->id, conn->id, session->nrec.name,
```

3. in iscsi_async_session_destruction(), cleaning up session and response the qtask. But it's hard to implement the response qtask with current code(no easy way to pass qtask from session_conn_shutdown to iscsi_async_session_destruction). 
It would be easy to do this based on your changes(iscsid: simplify qtask handling) where we can figure out the qtask easily. BTY, I really like this changes, it make the qtask handle clear and easy to understand, it can also fix the issue of failed to logout an unrecoverable session once iscsid is restarted.
So when would you make your change merged?

Reply:
That sounds good. Just take my patch

https://github.com/open-iscsi/open-iscsi/pull/292/commits/0795abbadef865bf3338f127ab03c990c8690422

and integrate into your patchset so it all gets tested together.

Reply:
> That sounds good. Just take my patch
> 
> [0795abb](https://github.com/open-iscsi/open-iscsi/commit/0795abbadef865bf3338f127ab03c990c8690422)
> 
> and integrate into your patchset so it all gets tested together.

I would integrate the patch and update this PR in next days.